Animal form

pablo-grande opened the pull request:
Beware travelers! here be dragons.

A few things to keep in mind before merging this branch:
1. The unknown breed is supposed to be existing in DB (ref: #1)
https://github.com/petconnection/backend/blob/9723c6c3ed820546202d770c9c2e2372cd250404/backoffice/views.py#L26
2. The form is ugly AF, needs a few upgrades (ref: #17)
3. You may have some troubles to enter and see the changes in local because of this:
https://github.com/petconnection/backend/blob/9723c6c3ed820546202d770c9c2e2372cd250404/backoffice/views.py#L9
We do not have a proper login yet (ref: #14) so... yeah, proably it can be bypassed by accessing first to `/admin` and logging there and then go to "View Site"

Commit:
Relocate statics folders

Commit:
Added new properties to Animal model

Property methods added:
    * Medical Record
    * Species

Returns both as if was a property
```
animal.species
```

Commit:
remove home and simply use / as backoffice home

Commit:
Add animal creation form

Create templates for animal form. Modify views.py to handle
request.methods for a POST method.
Create forms.py to animal form validation
Refactor templates to use includes for forms and other important
sections of HTML

pablo-grande commented:
4. About error messages on form submitting
https://github.com/petconnection/backend/blob/9723c6c3ed820546202d770c9c2e2372cd250404/backoffice/static/scripts.js#L27
This just gets a message provided by
https://github.com/petconnection/backend/blob/9723c6c3ed820546202d770c9c2e2372cd250404/backoffice/views.py#L53
Providing only hints if the form has some kind of validation error. This means if an exception occurs it would return nothing, so maybe we should keep that in mind and think a way of tackle this issue

Nosk14 commented on backoffice/views.py:
```diff
@@ -1,12 +1,55 @@
 from django.shortcuts import render
 from django.contrib.auth.decorators import login_required
-from api.models import Animal, Entity
+from api.models import *
```
Wildcard imports should be avoided. 
https://www.python.org/dev/peps/pep-0008/#imports

Commit:
:cop: Prevent * calling for models

pablo-grande commented:
5. No image uploading of the animal was implemented (sorry). Updating #17

Nosk14 commented on backoffice/views.py:
```diff
@@ -1,12 +1,55 @@
 
 @login_required(login_url='login')
 def home(request):
+    from django.contrib.auth import authenticate
```
Why is this import here? And where is it used?

pablo-grande replied:
This is the remainings of an old bypass for the login method and to avoid a `@login_required`. It should (and will be) removed.

Commit:
Remove unnecesary verbosity in __str__(Animal)

Commit:
Remove unnecesary imports

Remove auth import from legacy login function.
Remove datetime import, the Animal model already handles this
functionality

Commit:
Merge branch 'backoffice' into animal_form

Commit:
Delete lateral nav bar

Commit:
New animal form

New approach to add an animal.
Create a new view for adding an animal, no longer embedded
Modify forms/animal.html, 3 levels approach
scripts.js:
    * No longer AJAX request
    * Add click listeners to buttons in animal form

Modify header.html, checks variable in order to show clear and save
buttons on form
Modify views.py, move add animal to a new function
Modify urls.py to add `/add` route
Added css exception for card headers

pablo-grande commented:
Hello there! :wave: 
There was a MAJOR refactor of the form, check out this beauty:
![1](https://user-images.githubusercontent.com/5110926/38277403-9a63d158-3798-11e8-8ea7-f41de62220ce.png)

- [x] Looks way better. There is now a 3 level form, with separated contextual information.
- [x] Upper right buttons toogle between views, if we are adding an animal they will change into:
![2](https://user-images.githubusercontent.com/5110926/38277417-a02ce8a4-3798-11e8-8e04-bbe9a7ad145b.png)

- [x] Better viewing of errors if form failed
- [x] Change species selection for a dropdown select: http://creativeit.github.io/getmdl-select/
- [x] Edit button on animal cards should load an edit view with same form with populated fields
- [x] "Home" title should also be dynamic, telling the user where is he at every moment, with a `{% block second_title %}` maybe?

Commit:
Change species selection for a dropdown select

Commit:
Better viewing of errors if form failed

pablo-grande commented:
The new push changed color scheme in order to highlight errors

Commit:
Closes #17. Add edit feature to animal form

Modify views.py in order to use `update_or_create()` and tell wether the
app is editing or adding a new animal.
Modify template form to allow reading an animal object at editing
Add new url to enable editing
Substitute "Home" for a Paw icon that redirects to home at header
template

pablo-grande commented:
Sorry for the double commiting, I tried to close the issue in a cool way but I misspelled the number, I force pushed but it seems it didn't work well. And the funny thing is it is not completed! :sweat: 

- [x] Pic upload

- [x] Weight!

Commit:
Add weight to animal form

Commit:
Added animal photo input to animal form

Commit:
Add animal pic as background (if there is)

pablo-grande commented:
- [x] Delete feature

Commit:
Closes #17 Animal form

Implement delete function
Add js behaviour to prevent wrongful deletion
Add delete/{animal_id} route to urls.py

Commit:
Show delete button only in edit view

Commit:
Prevent animal pic deletion on edit

Add hidden input for animal pic in order to always get an animal pic on
editing, this is to prevent a the case a user edits an animal and do not changes the animal
photo input. The input will come as empty and hence updates the animal with no
picture at all.

Nosk14 commented on api/models.py:
```diff
@@ -43,10 +43,18 @@ class Animal(models.Model):
+        return "{} {}: {} {}".format(name, self.id, breed, species)
+
+    @property
+    def medical_record(self):
```
This property is not necessary. You can use `animal.medicalrecord` instead. It is an autogenerated property from the OneToOne relationship

Nosk14 commented on backoffice/static/scripts.js:
```diff
@@ -0,0 +1,64 @@
+
+
+  //asign semi-random color to a card
+  var colors = ['teal', 'pink', 'purple', 'deep-purpe', 'indigo', 'blue', 'light-blue',
```
There is a typo in `deep-purple`. The L is missing

Nosk14 reviewed (changes requested):
There are two strange behaviors in the clear button.
- Not all fields are cleared.
- After a clear, if I want to check a check button again, I need to click it twice.

Nosk14 committed:
typo fixed

Nosk14 committed:
removed property medical_record from animal model

Nosk14 committed:
removed clear button from animal form
